js/src/ReactView.js

```diff
@@ -0,0 +1,176 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import PropTypes from 'prop-types';
+import _ from 'lodash';
+import { styleWrapper } from './style_wrap';
+import { ReactWidgetModel } from './generated/ReactWidget';
+
+class WidgetComponent extends React.Component {
+    constructor(props) {
+        super(props);
+        this.el = React.createRef();
+    }
+
+    componentDidMount() {
+        const { model, view } = this.props;
+        const viewPromise = view.create_child_view(model);
+        viewPromise.then(childView => setTimeout(
+            () => JupyterPhosphorWidget.attach(childView.pWidget, this.el.current),
+            0,
+        ));
+    }
+
+    render() {
+        return React.createElement('div', { ref: this.el });
+    }
+}
+
+WidgetComponent.propTypes = {
+    model: PropTypes.PropTypes.instanceOf(WidgetModel).isRequired,
+    view: PropTypes.PropTypes.instanceOf(DOMWidgetView).isRequired,
+};
+
+class TopComponent extends React.Component {
+    componentWillUnmount() {
+        (this.allModels || []).forEach(model => model.off('change', this.updateCallback));
+    }
+
+    makeProps(model, ancestors) {
+        const snakeToCamel = (s) => {
+            let result = s.replace(/_$/, '');
+            if (result.startsWith('aria')) {
+                return result.replace('_', '-');
+            }
+            result = result.replace(/_\w/g, m => m[1].toUpperCase());
+            return result;
+        };
+
+        const hasAncestor = name => ancestors.map(m => m.name).some(n => n === name);
+
+        const { view } = this.props;
+
+        const childProps = model.keys()
+            .filter(k => !k.startsWith('_') && k !== 'layout' && model.get(k) != null)
+            .reduce((accumulator, key) => {
+                let v;
+                if (key === 'anchor_el') {
+                    v = document.querySelector(`[cid=${view.cid}${model.get(key).cid}]`);
+                } else if (key === 'value') { // value can be a widgetModel, we don't want to convert it to a react component
+                    v = model.get(key);
+                } else {
+                    v = this.convertModels(model.get(key), ancestors.concat(model));
+                }
+                return { ...accumulator, [snakeToCamel(key)]: v };
+            }, {});
+
+        if (model.keys().includes('checked')) {
+            const dontUseChecked = (model.name === 'RadioModel' && hasAncestor('RadioGroupModel'))
+                || (model.name === 'FormControlLabelModel' && hasAncestor('RadioGroupModel'));
+            if (!dontUseChecked) {
+                childProps.onChange = (e, checked) => {
+                    model.set('checked', checked);
+                    model.save_changes(model.callbacks(view));
+                };
+            }
+        } else if (model.keys().includes('selected')) {
+            const dontUseSelected = model.name === 'ToggleButtonModel' && hasAncestor('ToggleButtonGroupModel');
```

Comment:
some explanation would help

Reply:
double negation is difficult to read.
```suggestion
            const useSelected = !(model.name === 'ToggleButtonModel' && hasAncestor('ToggleButtonGroupModel'));
```